### ISSUE TITLE
intermittent hooks errors

### ISSUE THREAD

**Issue by hexsprite** [meteor/react-packages#297]
I've been getting intermittent hooks errors when using `useTracker`.

Not sure if the issue is somewhere in my own code (usually is haha) or if I uncovered an issue in this package.

The traceback is a little inconsistent and confusing as well. In one traceback it refers to ForwarfRef, in the 2nd it refers to UserMenu (one of the project's components which uses useTracker).

It doesn't happen all the time so it appears that it may be some kind of race condition...

![image](https://user-images.githubusercontent.com/23538/94858530-07050e80-03e8-11eb-8646-970eedf3618e.png)

**Comment by hexsprite** [comment]
Also note the Error from tracker recompute: at the bottom. It seems to indicate that the hook is getting CALLED from a Tracker recompute which seems sketchy... i have no idea why that would be

**Comment by hexsprite** [comment]
ps the referenced `UserMenu` is very dumb, doesn't use any other hooks other than UserMenu and otherwise renders simple HTML.

**Comment by CaptainN** [comment]
Can you show how you are using the hook?

Also, what version of React is that?

I see that you are not using `deps`, so that's maybe something to note.

**Comment by hexsprite** [comment]
```
  const { showSubscribe, showProUpgrade } = useTracker(() => {
    let showSubscribe = false
    let showProUpgrade = false
    const user = Meteor.user({ fields: { billing: 1 } })
    if (!Meteor.isCordova && user?.billing) {
      const { billing } = user
      showSubscribe = App.isTrialing() || App.isUnpaid()
      showProUpgrade = !showSubscribe && billing.plan !== 'pro'
    }
    return { showSubscribe, showProUpgrade }
  })
```

"react": "^16.13.1",

Re: deps - The only reason to change is if some user attribute changed so I didn't think it would need something in the dep part of the function.

**Comment by CaptainN** [comment]
I agree that you don't need to specify any deps values, but the presence of even an empty deps array does change the behavior of the hook. It might even solve your problem.

However, I'd like to make sure there's not a bug in the hook. There should be no change in the order of internal hook calls within the `useTracker` hook, unless something screwy happens, like the deps arg swaps between an array and undefined or something like that - but that's not happening here based on your example code.

Try setting `deps` to an empty array and see if that clears the problem. If it does, it might mean there's a bug in the `useTracker` implementation. If it doesn't, it still might be a bug, but it could also be in your code. (I don't think we are using ForwardRef for example)

**Comment by CaptainN** [comment]
What's going on inside `App.isTrialing() || App.isUnpaid()` - those aren't calling react hooks right?

**Comment by hexsprite** [comment]
No more usages of hooks.

They just retrieve Meteor.user() and look at some data on it.

So if they throw an error it will get swallowed?

**Comment by CaptainN** [comment]
It shouldn't swallow the error, but I'm honestly not sure. We should get a unit test or two for that.

I don't see anything in here to explain the error you posted. What version of `react-meteor-data` are you running?

**Comment by hexsprite** [comment]
Maybe I'll try throwing an error in there just to see what happens.

react-meteor-data@2.1.2

I noticed a couple other usages of useTracker and some of them should have had some deps actually. And I also changed the ones without deps to use an empty array. I'll followup if this comes back.

**Comment by hexsprite** [comment]
Still having this since I switched to use deps on all my usages of useTracker. I did notice on one of the tracebacks that it still says NoDeps... 

I did a search of some logs and I see that `withTracker` uses `useTracker` with the nodeps variant. So now I'm looking over my usages of that.

**Comment by CaptainN** [comment]
Yes, withTracker does use useTracker - that's why it's important for the nodeps path to work correctly.

If you could, please also check for cases where you might be calling a hook inside the computation itself.

**Comment by hexsprite** [comment]
continuing to have issues.

Note the latest traceback shows an internal React error that I'm curious about.

Having poured over my application code I'm relatively certain that it has nothing to do with breaking the rules of hooks.

It seems to get triggered when a reactive re-rendering occurs.

![image](https://user-images.githubusercontent.com/23538/97623460-eb8b2480-19e2-11eb-8585-c6d810e501fb.png)

I did this little video walk through that may shed some light. Any ideas on direction to go for debugging this?

https://www.loom.com/share/0901e081c07a4876bc0e309caf7d144e

**Comment by hexsprite** [comment]
Tried to create a repro using the blaze todos example app and adding a couple react components mounted using the react blaze helper. Wasn't able to trigger the bug. Now I'm gonna try turning off components one by one in the app until maybe I find something.

**Comment by hexsprite** [comment]
Looks like I may have found the issue. I was doing something like:

```
let EditComponent
if (something) {
  EditComponent = require('componentA').default
} else {
  EditComponent = require('componentA').default
}
return <EditComponent />
```

But apparently React did not like that!

**Comment by CaptainN** [comment]
You should definitely do your requires in the module scope, never in your render function.

**Comment by hexsprite** [comment]
That wasn't what I was doing before...

I had a Blaze Template like:

```
{{React component=component subcomponent=subcomponent}}
```

Then the react component was essentially:

```
function MyComponent (props) {
  const SubComponent = props.subcomponent;
  return <SubComponent />
}
```

So the subcomponent was passed as a prop to the main component

Not sure why that caused a (intermittent) React bug...

**Comment by CaptainN** [comment]
My guess is it was sending a different component (and render method) in some edge cases (maybe the component is being reuses for different routes, or something, and only changing the subcomponent).

It's still somewhat mysterious, but if it was solved this way, it's probably not a bug with useTracker. If you agree, please close this ticket.

I'm glad you found resolution. :-)
